Automattic/wp-calypso
# Jetpack Settings: Add actions, reducer, selectors and query components for connection status

### Introduction

This PR introduces actions, reducer and selectors for retrieving an actual Jetpack connection status. Also, a query component for managing network requests for it is being introduced. We will need these to be able to determine information about the current connection (like development mode or staging mode) in real time. This PR is part of the Jetpack Module Settings group effort (#9171).

### New WP.COM API methods
For interacting with the .com API it introduces a new method that is not yet implemented on the API side:

* `getJetpackConnectionStatus()`

In order to fully test the method you can:
* Apply D3588-code on your .com sandbox
* Apply https://github.com/Automattic/jetpack/pull/5418 on your Jetpack site (this PR is needed to allow remote requests to the Jetpack REST API).
* Connect your Jetpack site to your .com sandbox and verify they talk to each other.

### New actions and selectors

##### Actions

* `requestJetpackConnectionStatus( siteId )` - used to retrieve the current connection status

##### Selectors

* `isRequestingJetpackConnectionStatus` - whether the connection status is being requested currently
* `getJetpackConnectionStatus` - retrieve the current connection status
* `isJetpackSiteInDevelopmentMode` - whether the site is in development mode
* `isJetpackSiteInStagingMode` - whether the site is in staging mode

### To test

* Checkout this branch
* Play with the `QueryJetpackConnection` query component and verify it dispatches the proper network requests - `<QueryJetpackConnection siteId={ 1234 } />`.
* Verify all tests are working: `npm run test-client client/state/jetpack-settings/connection/`

/cc @oskosk @roccotripaldi @johnHackworth
